RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,538 @@
+                                        please call `miget` (or `fill_hadamard`) on your matrix before calling `invert_hadamard`");
+
+    // Creating a vector out of `selected_idxs` to be able to conveniently reverse
+    std::vector<sint_t> selected_idxs(hmo.selected_idxs, hmo.selected_idxs + hmo.d);
+    // Reversing the indices for performing the inverse
+    std::reverse(selected_idxs.begin(), selected_idxs.end());
+
+    //Step 1: Permute the rows/cols
+    if(hmc.left)
+        permute_rows_to_top(hmc.layout, hmc.n_rows, hmc.n_cols, selected_idxs.data(), hmo.d, hmc.A);
+    else
+        permute_cols_to_left(hmc.layout, hmc.n_rows, hmc.n_cols, selected_idxs.data(), hmo.d, hmc.A);
+
+    //Step 2: Apply the Hadamard transform (invH = H.T = H)
+    // This has to be a `double` because `blas::scal` really doesn't like being passed integer types
+    double padded_ld = std::pow(2, int(std::log2(hmo.dim)) + 1);
+    int64_t workspace_ld = (padded_ld - hmo.dim);
+
+    T log_sz = std::log2(hmo.dim);
+    T log_int_sz, log_final_sz;
+    T log_frac_sz = std::modf(log_sz, &log_int_sz);
+
+    if(log_frac_sz < 1e-3)
+        log_final_sz = log_int_sz;
+    else
+        log_final_sz = log_int_sz + 1;
+
+    fht_dispatch(hmc.left, hmc.layout, hmc.n_rows, hmc.n_cols, log_final_sz, hmc.A, workspace_ld, hmc.workspace);
+
+    // Scaling appropriately
+    blas::scal(hmc.n_rows * hmc.n_cols, 1 / padded_ld, hmc.A, 1);
+
+    //Step 3: Scale with `D` (invD = D for rademacher entries)
+    apply_diagonal_rademacher(hmc.left, hmc.layout, hmc.n_rows, hmc.n_cols, hmc.A, hmo.diag_scale);
+}
+
+/*
+ * Helper function to compute the size of the scratch workspace required
+ * Note, that this follows the same convention as `cuSolver` where the library provides
+ * helper functions for computing the size of scratch workspace required for a driver, often with
+ * a very similar signature to the driver itself
+*/
+template <typename T, typename state_t = RNGState<DefaultRNG>, SignedInteger sint_t = int64_t>
+inline int64_t miget_workspace_sz(
+    HadamardMixingOp<T, sint_t, state_t> &hmo,
+    HadamardMatrixCouple<T> &hmc) {
+    int64_t stride = hmc.left ? hmc.n_cols : hmc.n_rows;
+
+    // Grabs the power of 2 that is just bigger than the leading dimension
+    int64_t padded_ld = std::pow(2, int(std::log2(hmo.dim)) + 1);
+
+    return (padded_ld - hmo.dim) * stride;
+}
+
+/*
+ * Applies an in-place, SRHT-like transform to the input matrix
+ * i.e. A <- (\Pi H D)A OR A <- A(D H \Pi) (which is equivalent to A <- A(\Pi H D)^{-1})
+ * `HadamardMixingOp hmo`:  Data-oblivious description of the \PiRHT at hand
+ * `HadamardMatrixCouple hmc`: Description of the data to be sketched
+ */
+template <typename T, typename state_t = RNGState<DefaultRNG>, SignedInteger sint_t = int64_t>
+inline void miget(
+    HadamardMixingOp<T, sint_t> &hmo, // All information about `A` && the $\mathbb{\Pi\text{RHT}}$
```

Comment:
``// All information about `A` && the $\mathbb{\Pi\text{RHT}}$``

^ That's no longer true. Data for ``A`` isn't stored in HMO.